c2corg_api/views/user.py

```diff
@@ -0,0 +1,65 @@
+from functools import partial
+
+from cornice.resource import resource, view
+
+from c2corg_api.models.user import User, schema_user, schema_create_user
+from c2corg_api.views import (
+        cors_policy, json_view, to_json_dict)
+from c2corg_api.views.validation import validate_id
+
+from c2corg_api.models import DBSession
+
+
+def validate_unique_attribute(attrname, request):
+    """Checks if the given attribute is unique.
+    """
+
+    if attrname in request.json:
+        value = request.json[attrname]
+        attr = getattr(User, attrname)
+        count = DBSession.query(User).filter(attr == value).count()
+        if count == 0:
+            request.validated[attrname] = value
+        else:
+            request.errors.add('user', attrname, 'already used ' + attrname)
+
+
+@resource(path='/users/{id}', cors_policy=cors_policy)
+class UserRest(object):
+    def __init__(self, request):
+        self.request = request
+
+    @view(validators=validate_id)
+    def get(self):
+        id = self.request.validated['id']
+        user = DBSession. \
+            query(User). \
+            filter(User.id == id). \
+            first()
+
+        return to_json_dict(user, schema_user)
+
+
+@resource(path='/users/register', cors_policy=cors_policy)
+class UserRegistrationRest(object):
+    def __init__(self, request):
+        self.request = request
+
+    @json_view(schema=schema_create_user, validators=[
+        partial(validate_unique_attribute, "email"),
+        partial(validate_unique_attribute, "username")])
+    def post(self):
+        user = schema_create_user.objectify(self.request.validated)
+        # We need to encode the unicode string to UTF-8 to avoid
+        # "Unicode-objects must be encoded before hashing" error
+        user.password = self.request.json['password'].encode('utf-8')
+
+        DBSession.add(user)
+        try:
+            DBSession.flush()
+        except:
+            # Catching the exception ensures a json is sent back to the user
+            # and a message is shown.
+            self.request.errors.add('user', 'error', 'an error occured')
```

Comment:
But I am wondering: If you simply add something to this `errors` list (outside the validation), does this return the error messages? Because one line below you are returning the user as if everything worked fine.
I think you would have to raise a `BadRequest` like here: https://github.com/c2corg/v6_api/blob/master/c2corg_api/views/document.py#L321-L322